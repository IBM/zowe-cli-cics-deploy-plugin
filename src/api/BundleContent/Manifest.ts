/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright IBM Corp, 2019
*
*/

"use strict";

import { BundlePart, IBundlePartDataType } from "./BundlePart";
import { IHandlerParameters } from "@zowe/imperative";

/**
 * Interface to represent the manifest data for a CICS Bundle.
 *
 * @export
 * @interface IManifestType
 */
interface IManifestType {
  manifest: IManifestContents;
}
/**
 * Interface to represent the manifest data contents for CICS Bundle.
 *
 * @export
 * @interface IManifestContents
 */
interface IManifestContents {
  id?: string;
  xmlns?: string;
  bundleVersion?: number;
  bundleRelease?: number;
  bundleMajorVer?: number;
  bundleMinorVer?: number;
  bundleMicroVer?: number;
  define?: IBundlePartDataType[];
}

/**
 * Class to represent a CICS Bundle manifest.
 *
 * @export
 * @class Manifest
 */
export class Manifest {

  private MAX_BUNDLEID_LEN = 64;
  private fs = require("fs");
  private path = require("path");
  private manifestAsJson: IManifestType = undefined;
  private bundleDirectory: string;
  private metainfDir: string;
  private manifestFile: string;
  private manifestExists: boolean = false;
  private merge: boolean;
  private overwrite: boolean;
  private params: IHandlerParameters;
  private manifestAction = "    create";


  /**
   * Constructor for creating a Manifest.
   *
   * @static
   * @param {string} directory - The bundle directory.
   * @param {boolean} merge - Changes to the bundle manifest should be merged into any existing manifest.
   * @param {boolean} overwrite - Changes to the bundle contents should replace any existing contents.
   * @param {IHandlerParameters} params - The current Imperative handler parameters
   * @throws ImperativeError
   * @memberof Manifest
   */
  constructor(directory: string, merge: boolean, overwrite: boolean, params?: IHandlerParameters) {
   this.merge = merge;
   this.overwrite = overwrite;
   this.bundleDirectory = this.path.normalize(directory);
   this.metainfDir = this.bundleDirectory + "/META-INF";
   this.manifestFile = this.metainfDir + "/cics.xml";
   this.params = params;

   // If 'merge' is set then attempt to read any existing manifest that may
   // already exist. Subsequent changes will be merged with the existing
   // content.
   if (this.merge === true) {
     try {
       this.readManifest();
     } catch (ex) {
       // Something went wrong. If it's an ENOENT response then there was
       // no manifest to read, so that can be ignored, otherwise propagate
       // the exception back to the caller.
       if (ex.code !== "ENOENT") {
         throw ex;
       }
     }
   }

   // If we've not read an existing manifest, create a new one
   if (this.manifestAsJson === undefined) {
     this.manifestAsJson = { manifest:
                             { xmlns: "http://www.ibm.com/xmlns/prod/cics/bundle",
                               bundleVersion: 1,
                               bundleRelease: 0 } };
   }
  }

  /**
   * Perform whatever validation can be done in advance of attempting to save the
   * manifest, thereby reducing the possibility of a failure after some of the
   * bundle parts have already been persisted to the file system.
   *
   * @throws ImperativeError
   * @memberof Manifest
   */
  public prepareForSave() {
    // Does the meta-inf directory already exist?
    if (!this.fs.existsSync(this.metainfDir)) {
      // we'll have to create it during the save, do we have write permission?
      try {
        this.fs.accessSync(this.bundleDirectory, this.fs.constants.W_OK);
      }
      catch (err) {
        throw new Error("cics-deploy requires write permission to: " + this.bundleDirectory);
      }

      return;
    }

    // Do we have write permission to the META-INF dir?
    try {
      this.fs.accessSync(this.metainfDir, this.fs.constants.W_OK);
    }
    catch (err) {
      throw new Error("cics-deploy requires write permission to: " + this.metainfDir);
    }

    // Does a manifest file already exist?
    if (this.fs.existsSync(this.manifestFile)) {
      if (this.overwrite === false) {
        throw new Error("A bundle manifest file already exists. Specify --overwrite to replace it, or --merge to merge changes into it.");
      }
      if (this.merge) {
        this.manifestAction = "     merge";
      }
      else {
        this.manifestAction = " overwrite";
      }

      // Do we have write permission to the manifest?
      try {
        this.fs.accessSync(this.manifestFile, this.fs.constants.W_OK);
      }
      catch (err) {
        throw new Error("cics-deploy requires write permission to: " + this.manifestFile);
      }
    }
  }

  /**
   * Save the Manifest file.
   *
   * @throws ImperativeError
   * @memberof Manifest
   */
  public save() {
     // Create the META-INF directory if it doesn't already exist
     if (!this.fs.existsSync(this.metainfDir)) {
       try {
         this.logCreation(this.metainfDir);
         this.fs.mkdirSync(this.metainfDir);
       }
       catch (err) {
         throw new Error("An error occurred attempting to create '" + this.metainfDir + "': " + err.message);
       }
     }

     // Write the cics.xml manifest
     try {
       this.logCreation(this.manifestFile, this.manifestAction);
       this.fs.writeFileSync(this.manifestFile, this.getXML(), "utf8");
     }
     catch (err) {
       throw new Error("An error occurred attempting to write manifest file '" + this.manifestFile + "': " + err.message);
     }

     this.manifestExists = true;
  }

  /**
   * Validates that a manifest file exists. If the manifest has not yet been saved then it is
   * invalid.
   *
   * @returns {object}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public validate() {
    if (!this.manifestExists) {
      throw new Error("No bundle manifest file found: " + this.manifestFile);
    }
  }

  /**
   * Returns the Manifest contents as a JSON Object.
   *
   * @returns {object}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public getJson(): object {
    return this.manifestAsJson;
  }

  /**
   * Returns the Manifest contents as XML text.
   *
   * @returns {string}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public getXML(): string {
    const parser = require("xml2json");
    return parser.toXml(JSON.stringify(this.manifestAsJson)) + "\n";
  }

  /**
   * Returns the Bundle's identity (id) value.
   *
   * @returns {string}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public getBundleId(): string {
    return this.manifestAsJson.manifest.id;
  }

  /**
   * Set the Bundle's identity (id) value.
   * @param {string} id - The identiry value for the Bundle.
   * @throws ImperativeError
   * @memberof Manifest
   */
  public setBundleId(id: string) {
    if (id === undefined) {
      throw new Error("BundleId not set.");
    }
    const mangledId = this.mangle(id, this.MAX_BUNDLEID_LEN);
    this.manifestAsJson.manifest.id = mangledId;
  }

  /**
   * Returns the Bundle's version number.
   *
   * @returns {string}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public getBundleVersion(): string {
    if (this.manifestAsJson.manifest.bundleMajorVer === undefined) {
      this.manifestAsJson.manifest.bundleMajorVer = 1;
    }
    if (this.manifestAsJson.manifest.bundleMinorVer === undefined) {
      this.manifestAsJson.manifest.bundleMinorVer = 0;
    }
    if (this.manifestAsJson.manifest.bundleMicroVer === undefined) {
      this.manifestAsJson.manifest.bundleMicroVer = 0;
    }

    return this.manifestAsJson.manifest.bundleMajorVer + "." +
           this.manifestAsJson.manifest.bundleMinorVer + "." +
           this.manifestAsJson.manifest.bundleMicroVer;
  }

  /**
   * Set the Bundle's version number.
   * @param {number} majorVersion - The major version number.
   * @param {number} minorVersion - The minor version number.
   * @param {number} microVersion - The micro version number.
   * @throws ImperativeError
   * @memberof Manifest
   */
  public setBundleVersion(majorVersion: number, minorVersion: number, microVersion: number) {
    if (Number.isInteger(majorVersion) && majorVersion > 0) {
      this.manifestAsJson.manifest.bundleMajorVer = majorVersion;
    }
    else {
      throw new Error("Invalid Bundle major version specified: " + majorVersion + ".");
    }

    if (Number.isInteger(minorVersion) && minorVersion >= 0) {
      this.manifestAsJson.manifest.bundleMinorVer = minorVersion;
    }
    else {
      throw new Error("Invalid Bundle minor version specified: " + minorVersion + ".");
    }

    if (Number.isInteger(microVersion) && microVersion >= 0) {
      this.manifestAsJson.manifest.bundleMicroVer = microVersion;
    }
    else {
      throw new Error("Invalid Bundle micro version specified: " + microVersion + ".");
    }
  }

  /**
   * Add a resource definition to the Manifest. If a part of the same name and type
   * already exists then it is replaced with the new part data.
   *
   * @param {BundlePart} bundlePart - the BundlePart to add.
   * @throws ImperativeError
   * @memberof Manifest
   */
  public addDefinition(bundlePart: BundlePart) {

    const definition = bundlePart.getPartData();

    // Ensure that the definitions array is usable
    this.prepareDefinitionsArray();

    // Search the array for a definition with the same name and
    // type as the one we're trying to add, if found then update
    // it with the new path.
    let i: number;
    for (i = 0; i < this.manifestAsJson.manifest.define.length; i++) {
      const candidate = this.manifestAsJson.manifest.define[i];
      if (candidate.name === definition.name &&
          candidate.type === definition.type) {
        candidate.path = definition.path;
        return;
      }
    }

    // If we've not replaced an existing item in the list, add
    // the new item to the end.
    this.manifestAsJson.manifest.define.push(definition);
  }

  /**
   * Does the bundle contain any resource definitions of the
   * specified type?
   *
   * @param {String} resourceType - the resource type
   * @throws ImperativeError
   * @memberof Manifest
   */
  public containsDefinitionsOfType(resourceType: string): boolean {

    // Ensure that the definitions array is usable
    this.prepareDefinitionsArray();

    // Search the array for a definition with the nominated type
    let i: number;
    for (i = 0; i < this.manifestAsJson.manifest.define.length; i++) {
      const candidate = this.manifestAsJson.manifest.define[i];
      if (candidate.type === resourceType) {
        return true;
      }
    }

    return false;
  }


  private prepareDefinitionsArray() {
    // Create the array of definitions if it doesn't already exist
    if (this.manifestAsJson.manifest.define === undefined) {
      this.manifestAsJson.manifest.define = [];
    }

    // If what already exists is an object (as may happen when appending
    // to an existing XML manifest with only one entry) convert it to an array;
    if (Array.isArray(this.manifestAsJson.manifest.define) === false) {
      const obj = this.manifestAsJson.manifest.define as any;
      this.manifestAsJson.manifest.define = [];
      this.manifestAsJson.manifest.define.push(obj as IBundlePartDataType);
    }
  }

  // Read an existing manifest file into an object
  private readManifest() {
    // read the manifest from the file system
    const xmltext = this.fs.readFileSync(this.manifestFile, "utf8");

    try {
      // Reading the file worked, so convert the contents into a JSON Object
      const parser = require("xml2json");
      this.manifestAsJson = JSON.parse( parser.toJson( xmltext, { reversible: true } ) );
    }
    catch (exception)
    {
      throw new Error("Parsing error occurred reading a CICS manifest file: " + exception.message);
    }

    // Note, the XML parser has no support for namespaces.
    // If the manifest is namespace prefixed then it wont be found. Yuck.
    // Validate that the manifest we've read is usable.
    if (this.manifestAsJson.manifest === undefined) {
      throw new Error("Existing CICS Manifest file found with unparsable content.");
    }

    // Now check the namespace
    if (this.manifestAsJson.manifest.xmlns !== "http://www.ibm.com/xmlns/prod/cics/bundle") {
      throw new Error("Existing CICS Manifest file found with unexpected namespace: " + this.manifestAsJson.manifest.xmlns + " .");
    }

    this.manifestExists = true;
  }

  // Function for mangaling a string for inclusion in the manifest
  private mangle(text: string, maxLength: number): string {

    // replace underscore with hyphen
    const text2 = text.replace(/_/g, "-");

    // Convert all unsupported characters to X
    const text3 = text2.replace(/[^0-9,^A-Z,^a-z\-]/g, "X");

    // Now truncate the string
    const text4 = text3.substring(0, maxLength);

    return text4;
  }

  private logCreation(file: string, action?: string) {
    if (action === undefined) {
      action = "    create";
    }
    if (this.params !== undefined) {
      this.params.response.console.log(action + " : " + this.path.relative(this.bundleDirectory, file));
    }
  }
}
