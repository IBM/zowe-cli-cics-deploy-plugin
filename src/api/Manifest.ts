/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright IBM Corp, 2019
*
*/

"use strict";

import { BundlePart, IBundlePartDataType } from "./BundlePart";

/**
 * Interface to represent the manifest data for a CICS Bundle.
 *
 * @export
 * @interface IManifestType
 */
interface IManifestType {
  manifest: IManifestContents;
}
/**
 * Interface to represent the manifest data contents for CICS Bundle.
 *
 * @export
 * @interface IManifestContents
 */
interface IManifestContents {
  id?: string;
  xmlns?: string;
  bundleVersion?: number;
  bundleRelease?: number;
  bundleMajorVer?: number;
  bundleMinorVer?: number;
  bundleMicroVer?: number;
  define?: IBundlePartDataType[];
}

/**
 * Class to represent a CICS Bundle manifest.
 *
 * @export
 * @class Manifest
 */
export class Manifest {

  private MAX_BUNDLEID_LEN = 64;
  private fs = require("fs");
  private path = require("path");
  private manifestAsJson: IManifestType;
  private bundleDirectory: string;
  private metainfDir: string;
  private manifestFile: string;


  /**
   * Constructor for creating a Manifest.
   *
   * @static
   * @param {string} directory - The bundle directory.
   * @throws ImperativeError
   * @memberof Manifest
   */
  constructor(directory: string) {
   this.bundleDirectory = this.path.normalize(directory);
   this.metainfDir = this.bundleDirectory + "/META-INF";
   this.manifestFile = this.metainfDir + "/cics.xml";

   // If there is an existing manifest file, read it
   try {
     this.readManifest();
   } catch (ex) {
     if (ex.code === "ENOENT") {
      // No existing file was found, so magic up an empty manifest
       this.manifestAsJson = { manifest:
                               { xmlns: "http://www.ibm.com/xmlns/prod/cics/bundle",
                                 bundleVersion: 1,
                                 bundleRelease: 0 } };
       return;
     }
     throw ex;
   }
  }

  /**
   * Save the Manifest file.
   *
   * @throws ImperativeError
   * @memberof Manifest
   */
  // Save the manifest file back into the file system
  public save() {
     // Create the META-INF directory if it doesn't already exist
     if (!this.fs.existsSync(this.metainfDir)) {
      this.fs.mkdirSync(this.metainfDir);
     }

     // Write the cics.xml manifest
     this.fs.writeFileSync(this.manifestFile, this.getXML(), "utf8");
  }

  /**
   * Returns the Manifest contents as a JSON Object.
   *
   * @returns {object}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public getJson(): object {
    return this.manifestAsJson;
  }

  /**
   * Returns the Manifest contents as XML text.
   *
   * @returns {string}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public getXML(): string {
    const parser = require("xml2json");
    return parser.toXml(JSON.stringify(this.manifestAsJson)) + "\n";
  }

  /**
   * Returns the Bundle's identity (id) value.
   *
   * @returns {string}
   * @throws ImperativeError
   * @memberof Manifest
   */
  public getBundleId(): string {
    return this.manifestAsJson.manifest.id;
  }

  /**
   * Set the Bundle's identity (id) value.
   * @param {string} id - The identiry value for the Bundle.
   * @throws ImperativeError
   * @memberof Manifest
   */
  public setBundleId(id: string) {
    if (id === undefined) {
      throw new Error("BundleId not set.");
    }
    const mangledId = this.mangle(id, this.MAX_BUNDLEID_LEN);
    this.manifestAsJson.manifest.id = mangledId;
  }

  /**
   * Set the Bundle's version number.
   * @param {number} majorVersion - The major version number.
   * @param {number} minorVersion - The minor version number.
   * @param {number} microVersion - The micro version number.
   * @throws ImperativeError
   * @memberof Manifest
   */
  public setBundleVersion(majorVersion: number, minorVersion: number, microVersion: number) {
    if (Number.isInteger(majorVersion) && majorVersion > 0) {
      this.manifestAsJson.manifest.bundleMajorVer = majorVersion;
    }
    else {
      throw new Error("Invalid Bundle version specified.");
    }

    if (Number.isInteger(minorVersion) && minorVersion >= 0) {
      this.manifestAsJson.manifest.bundleMinorVer = minorVersion;
    }
    else {
      throw new Error("Invalid Bundle version specified.");
    }

    if (Number.isInteger(microVersion) && microVersion >= 0) {
      this.manifestAsJson.manifest.bundleMicroVer = microVersion;
    }
    else {
      throw new Error("Invalid Bundle version specified.");
    }
  }

  /**
   * Add a resource definition to the Manifest. If a part of the same name and type
   * already exists then it is replaced with the new part data.
   *
   * @param {BundlePart} bundlePart - the BundlePart to add.
   * @throws ImperativeError
   * @memberof Manifest
   */
  public addDefinition(bundlePart: BundlePart) {

    const definition = bundlePart.getPartData();

    // Create the array of definitions if it doesn't already exist
    if (this.manifestAsJson.manifest.define === undefined) {
      this.manifestAsJson.manifest.define = [];
    }

    // If what already exists is an object (as may happen when appending
    // to an existing XML manifest with only one entry) convert it to an array;
    if (Array.isArray(this.manifestAsJson.manifest.define) === false) {
      const obj = this.manifestAsJson.manifest.define as any;
      this.manifestAsJson.manifest.define = [];
      this.manifestAsJson.manifest.define.push(obj as IBundlePartDataType);
    }

    // Search the array for a definition with the same name and
    // type as the one we're trying to add, if found then update
    // it with the new path.
    let i: number;
    for (i = 0; i < this.manifestAsJson.manifest.define.length; i++) {
      const candidate = this.manifestAsJson.manifest.define[i];
      if (candidate.name === definition.name &&
          candidate.type === definition.type) {
        candidate.path = definition.path;
        return;
      }
    }

    // If we've not replaced an existing item in the list, add
    // the new item to the end.
    this.manifestAsJson.manifest.define.push(definition);
  }

  // Read an existing manifest file into an object
  private readManifest() {
    // read the manifest from the file system
    const xmltext = this.fs.readFileSync(this.manifestFile, "utf8");

    try {
      // Reading the file worked, so convert the contents into a JSON Object
      const parser = require("xml2json");
      this.manifestAsJson = JSON.parse( parser.toJson( xmltext, { reversible: true } ) );
    }
    catch (exception)
    {
      throw new Error("Parsing error occurred reading a CICS manifest file: " + exception.message);
    }

    // Note, the XML parser has no support for namespaces.
    // If the manifest is namespace prefixed then it wont be found. Yuck.
    // Validate that the manifest we've read is usable.
    if (this.manifestAsJson.manifest === undefined) {
      throw new Error("Existing CICS Manifest file found with unparsable content.");
    }

    // Now check the namespace
    if (this.manifestAsJson.manifest.xmlns !== "http://www.ibm.com/xmlns/prod/cics/bundle") {
      throw new Error("Existing CICS Manifest file found with unexpected namespace: " + this.manifestAsJson.manifest.xmlns + " .");
    }
  }

  // Function for mangaling a string for inclusion in the manifest
  private mangle(text: string, maxLength: number): string {

    // replace underscore with hyphen
    const text2 = text.replace(/_/g, "-");

    // Convert all unsupported characters to X
    const text3 = text2.replace(/[^0-9,^A-Z,^a-z\-]/g, "X");

    // Now truncate the string
    const text4 = text3.substring(0, maxLength);

    return text4;
  }
}
