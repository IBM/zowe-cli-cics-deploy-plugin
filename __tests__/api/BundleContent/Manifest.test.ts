/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright IBM Corp, 2019
*
*/

import { Manifest } from "../../../src/api/BundleContent/Manifest";
describe("Manifest01", () => {
    it("should read an existing manifest", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/ExampleBundle01", true, true);

        // Check the output as JSON
        expect(JSON.stringify(man.getJson())).toMatchSnapshot();
        // Check the output as XML
        expect(man.getXML()).toMatchSnapshot();
    });
    it("should create an empty manifest from empty dir", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/EmptyBundle01", true, true);

        // Check the output
        expect(JSON.stringify(man.getJson())).toMatchSnapshot();
    });
    it("should create an empty manifest from empty META-INF dir", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/EmptyBundle02", true, true);

        // Check the output
        expect(JSON.stringify(man.getJson())).toMatchSnapshot();
    });
    it("set the bundleId", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/ExampleBundle01", true, true);

        // Set a bundleId
        man.setBundleId("testing a BundleId");

        // Check the output as JSON
        expect(JSON.stringify(man.getJson())).toMatchSnapshot();
        expect(man.getBundleId()).toMatchSnapshot();
    });
    it("set a long  bundleId", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/ExampleBundle01", true, true);

        // Set a bundleId
        man.setBundleId("1234567890123456789012345678901234567890123456789012345678901234567890");

        // Check the output as JSON
        expect(JSON.stringify(man.getJson())).toMatchSnapshot();
        expect(man.getBundleId()).toMatchSnapshot();
    });
    it("set a null bundleId", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/ExampleBundle01", true, true);

        // Add a definition
        let err: Error;
        try {
          man.setBundleId(undefined);
        }
        catch (error) {
          err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("set a valid version number", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/ExampleBundle01", true, true);

        // Set a bundleId
        man.setBundleVersion(1, 2, 3);

        // Check the output as JSON
        expect(JSON.stringify(man.getJson())).toMatchSnapshot();
    });
    it("set an invalid major version number", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/EmptyBundle01", true, true);

        let err: Error;
        try {
          // Set a bundleId
          man.setBundleVersion(-1, 0, 0);
        }
        catch (error) {
          err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("set an invalid minor version number", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/EmptyBundle01", true, true);

        let err: Error;
        try {
          // Set a bundleId
          man.setBundleVersion(1, 3.4, 0);
        }
        catch (error) {
          err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("set an invalid micro version number", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/EmptyBundle01", true, true);

        let err: Error;
        try {
          // Set a bundleId
          man.setBundleVersion(1, 3, 4.3);
        }
        catch (error) {
          err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("set a null version number", () => {

        // Create a Manifest
        const man = new Manifest("__tests__/__resources__/EmptyBundle01", true, true);

        let err: Error;
        try {
          // Set a bundleId
          man.setBundleVersion(undefined, undefined, undefined);
        }
        catch (error) {
          err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("Parse a manifest with bad namespace", () => {

        // Read a bad manifest
        let err: Error;
        try {
          const man = new Manifest("__tests__/__resources__/BadManifestBundle02", true, true);
        }
        catch (error) {
          err = error;
        }

        // Check the output as JSON
        expect(err.message).toContain("Existing CICS Manifest file found with unexpected namespace: wibble .");
    });
    it("Parse a malformed manifest", () => {

        // Read a bad manifest
        let err: Error;
        try {
          const man = new Manifest("__tests__/__resources__/BadManifestBundle03", true, true);
        }
        catch (error) {
          err = error;
        }

        // Check the output as JSON
        expect(err.message).toContain("Existing CICS Manifest file found with unparsable content:");
    });
});
