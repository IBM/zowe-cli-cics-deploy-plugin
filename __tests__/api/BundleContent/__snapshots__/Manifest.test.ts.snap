// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manifest01 set a long  bundleId 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"1234567890123456789012345678901234567890123456789012345678901234","bundleMajorVer":"10","bundleMinorVer":"11","bundleMicroVer":"12","define":[{"name":"name1","type":"type1","path":"path1"},{"name":"name2","type":"type2","path":"path2"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"}]}}"`;

exports[`Manifest01 set a long  bundleId 2`] = `"1234567890123456789012345678901234567890123456789012345678901234"`;

exports[`Manifest01 set a null bundleId 1`] = `"BundleId not set."`;

exports[`Manifest01 set a null version number 1`] = `"Invalid Bundle major version specified: undefined."`;

exports[`Manifest01 set a valid version number 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"ThisIsAnId","bundleMajorVer":1,"bundleMinorVer":2,"bundleMicroVer":3,"define":[{"name":"name1","type":"type1","path":"path1"},{"name":"name2","type":"type2","path":"path2"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"}]}}"`;

exports[`Manifest01 set an invalid major version number 1`] = `"Invalid Bundle major version specified: -1."`;

exports[`Manifest01 set an invalid micro version number 1`] = `"Invalid Bundle micro version specified: 4.3."`;

exports[`Manifest01 set an invalid minor version number 1`] = `"Invalid Bundle minor version specified: 3.4."`;

exports[`Manifest01 set the bundleId 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"testingXaXBundleId","bundleMajorVer":"10","bundleMinorVer":"11","bundleMicroVer":"12","define":[{"name":"name1","type":"type1","path":"path1"},{"name":"name2","type":"type2","path":"path2"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"}]}}"`;

exports[`Manifest01 set the bundleId 2`] = `"testingXaXBundleId"`;

exports[`Manifest01 should create an empty manifest from empty META-INF dir 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":1,"bundleRelease":0}}"`;

exports[`Manifest01 should create an empty manifest from empty dir 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":1,"bundleRelease":0}}"`;

exports[`Manifest01 should read an existing manifest 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"ThisIsAnId","bundleMajorVer":"10","bundleMinorVer":"11","bundleMicroVer":"12","define":[{"name":"name1","type":"type1","path":"path1"},{"name":"name2","type":"type2","path":"path2"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"}]}}"`;

exports[`Manifest01 should read an existing manifest 2`] = `
"<manifest xmlns="http://www.ibm.com/xmlns/prod/cics/bundle" bundleVersion="1" bundleRelease="2" id="ThisIsAnId" bundleMajorVer="10" bundleMinorVer="11" bundleMicroVer="12"><define name="name1" type="type1" path="path1"></define><define name="name2" type="type2" path="path2"></define><define name="name3" type="http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP" path="nodejsapps/Test.nodejsapp"></define></manifest>
"
`;
